sentry-apollo-3/src/main/java/io/sentry/apollo3/SentryApollo3HttpInterceptor.kt

```diff
@@ -0,0 +1,162 @@
+package io.sentry.apollo3
+
+import com.apollographql.apollo3.api.http.HttpRequest
+import com.apollographql.apollo3.api.http.HttpResponse
+import com.apollographql.apollo3.exception.ApolloNetworkException
+import com.apollographql.apollo3.network.http.HttpInterceptor
+import com.apollographql.apollo3.network.http.HttpInterceptorChain
+import io.sentry.Breadcrumb
+import io.sentry.Hint
+import io.sentry.HubAdapter
+import io.sentry.IHub
+import io.sentry.ISpan
+import io.sentry.SentryLevel
+import io.sentry.SpanStatus
+import io.sentry.TracingOrigins
+import io.sentry.TypeCheckHint
+import okio.Buffer
+import java.io.InputStreamReader
+
+class SentryApollo3HttpInterceptor(private val hub: IHub = HubAdapter.getInstance(), private val beforeSpan: BeforeSpanCallback? = null) :
+    HttpInterceptor {
+
+    override suspend fun intercept(
+        request: HttpRequest,
+        chain: HttpInterceptorChain
+    ): HttpResponse {
+        val activeSpan = hub.span
+        return if (activeSpan == null) {
+            chain.proceed(request)
+        } else {
+            val span = startChild(request, activeSpan)
+
+            val requestBuilder = request.newBuilder()
+
+            if (TracingOrigins.contain(hub.options.tracingOrigins, request.url)) {
+                val sentryTraceHeader = span.toSentryTrace()
+                val baggageHeader = span.toBaggageHeader()
+                requestBuilder.addHeader(sentryTraceHeader.name, sentryTraceHeader.value)
+
+                baggageHeader?.let {
+                    requestBuilder.addHeader(it.name, it.value)
+                }
+            }
+
+            val modifiedRequest = requestBuilder.build()
+
+            return try {
+                val httpResponse = chain.proceed(modifiedRequest)
+                span.status = SpanStatus.fromHttpStatusCode(httpResponse.statusCode, SpanStatus.UNKNOWN)
+                finish(span, modifiedRequest, httpResponse)
+
+                httpResponse
+            } catch (e: Exception) {
+                when (e) {
+                    is ApolloNetworkException -> span.status = SpanStatus.INTERNAL_ERROR
+                    else -> SpanStatus.UNKNOWN
+                }
+                finish(span, modifiedRequest)
+                throw e
+            }
+        }
+    }
+
+    private fun Long?.ifHasValidLength(fn: (Long) -> Unit) {
+        if (this != null && this != -1L) {
+            fn.invoke(this)
+        }
+    }
+
+    private fun startChild(request: HttpRequest, activeSpan: ISpan): ISpan {
+        val buffer = Buffer()
+        request.body?.writeTo(buffer)
+
+        val reader = InputStreamReader(buffer.inputStream())
+
+        val content = try {
+            hub.options.serializer.deserialize(
+                reader,
+                ApolloRequestBodyContent::class.java,
+                ApolloRequestBodyContent.Deserializer
+            )
```

Comment:
Maybe we just copy the body array and let the serialization happen during the event processing which is in a background thread? otherwise, this would affect the app's performance in the calling thread, it could be the main thread, `CachedBodyHttpServletRequest` does something similar I guess.

Reply:
By default the calling thread is a Background thread, ApolloClient uses Dispatcher.IO, unless explicitly defined otherwise.
The calling thread itself, however, is blocked, so it I'll have a look at using an `EventProcessor`

Reply:
Nope, that's not what I meant, unless this depends on `request.body`, if the graph query and operation are in the `body`, then we need it.
`ApolloInterceptor` offers access such as `request.operation.name()` but not sure if there's everything either.
I guess you both are on top of this, I was just wondering if the deserialization was necessary, if it is, well... nothing to do.